File: packages/content/src/Models/Post.php
<?php

declare(strict_types=1);

namespace Plog\Content\Models;

use Plog\Db\Connection\ConnectionInterface;

/**
 * 文章模型
 */
class Post
{
    /**
     * 数据库连接
     *
     * @var ConnectionInterface
     */
    private ConnectionInterface $connection;

    /**
     * 文章 ID
     *
     * @var int
     */
    public int $gid;

    /**
     * 文章标题
     *
     * @var string
     */
    public string $title;

    /**
     * 文章别名
     *
     * @var string
     */
    public string $slug;

    /**
     * 文章内容
     *
     * @var string
     */
    public string $content;

    /**
     * 文章摘要
     *
     * @var string
     */
    public string $excerpt = '';

    /**
     * 作者 ID
     *
     * @var int
     */
    public int $author;

    /**
     * 分类 ID
     *
     * @var int
     */
    public int $sortid = 0;

    /**
     * 发布时间
     *
     * @var string
     */
    public string $date;

    /**
     * 隐藏状态
     *
     * @var string
     */
    public string $hide = 'n';

    /**
     * 允许评论
     *
     * @var string
     */
    public string $allow_remark = 'y';

    /**
     * 置顶
     *
     * @var string
     */
    public string $top = 'n';

    /**
     * 类型
     *
     * @var string
     */
    public string $type = 'blog';

    /**
     * 密码
     *
     * @var string
     */
    public string $password = '';

    /**
     * 文章状态常量
     */
    const STATUS_DRAFT = 'draft';
    const STATUS_PUBLISHED = 'n';
    const STATUS_HIDDEN = 'y';

    /**
     * 文章类型常量
     */
    const TYPE_POST = 'blog';
    const TYPE_PAGE = 'page';

    /**
     * 构造函数
     *
     * @param ConnectionInterface $connection 数据库连接
     */
    public function __construct(ConnectionInterface $connection)
    {
        $this->connection = $connection;
    }

    /**
     * 获取所有文章
     *
     * @return array
     */
    public function all(): array
    {
        return $this->connection->table('blog')
            ->where('type', '=', self::TYPE_POST)
            ->orderBy('date', 'DESC')
            ->get();
    }

    /**
     * 获取已发布的文章
     *
     * @return array
     */
    public function published(): array
    {
        return $this->connection->table('blog')
            ->where('hide', '=', 'n')
            ->where('type', '=', self::TYPE_POST)
            ->orderBy('date', 'DESC')
            ->get();
    }

    /**
     * 根据主键查找
     *
     * @param int $id 文章 ID
     * @return array|null
     */
    public function find(int $id): ?array
    {
        return $this->connection->table('blog')
            ->where('gid', '=', $id)
            ->first();
    }

    /**
     * 根据别名查找
     *
     * @param string $slug 文章别名
     * @return array|null
     */
    public function findBySlug(string $slug): ?array
    {
        return $this->connection->table('blog')
            ->where('slug', '=', $slug)
            ->first();
    }

    /**
     * 创建文章
     *
     * @param array $data 文章数据
     * @return int 文章 ID
     */
    public function create(array $data): int
    {
        $data['date'] = $data['date'] ?? date('Y-m-d H:i:s');
        $data['type'] = $data['type'] ?? self::TYPE_POST;

        return $this->connection->table('blog')->insert($data);
    }

    /**
     * 更新文章
     *
     * @param int $id 文章 ID
     * @param array $data 文章数据
     * @return int 影响的行数
     */
    public function update(int $id, array $data): int
    {
        return $this->connection->table('blog')
            ->where('gid', '=', $id)
            ->update($data);
    }

    /**
     * 删除文章
     *
     * @param int $id 文章 ID
     * @return int 影响的行数
     */
    public function delete(int $id): int
    {
        return $this->connection->table('blog')
            ->where('gid', '=', $id)
            ->delete();
    }

    /**
     * 分页获取文章
     *
     * @param int $page 页码
     * @param int $perPage 每页数量
     * @return array
     */
    public function paginate(int $page = 1, int $perPage = 20): array
    {
        $offset = ($page - 1) * $perPage;

        return $this->connection->table('blog')
            ->where('type', '=', self::TYPE_POST)
            ->orderBy('date', 'DESC')
            ->limit($perPage)
            ->offset($offset)
            ->get();
    }
}

File: plog-rs/modules/content/src/repository/post.rs
//! Post Repository

use sea_orm::{*, prelude::Expr};
use std::sync::Arc;

use crate::entities::post::*;
use plog_shared::apply_if_set;

pub struct PostRepository {
    db: Arc<DatabaseConnection>,
}

impl PostRepository {
    pub fn new(db: Arc<DatabaseConnection>) -> Self {
        Self { db }
    }

    pub fn db(&self) -> &Arc<DatabaseConnection> {
        &self.db
    }

    pub async fn find_by_id(&self, id: i32) -> Result<Option<Model>, DbErr> {
        Entity::find_by_id(id).one(&*self.db).await
    }

    pub async fn find_by_alias(&self, alias: &str) -> Result<Option<Model>, DbErr> {
        Entity::find()
            .filter(Column::Alias.eq(alias))
            .one(&*self.db)
            .await
    }

    pub async fn paginate(&self, page: u64, per_page: u64) -> Result<(Vec<Model>, u64), DbErr> {
        let paginator = Entity::find()
            .order_by_desc(Column::Date)
            .paginate(&*self.db, per_page);
        let total = paginator.num_items().await?;
        let items = paginator.fetch_page(page.saturating_sub(1)).await?;
        Ok((items, total))
    }

    pub async fn find_published(&self, page: u64, per_page: u64) -> Result<(Vec<Model>, u64), DbErr> {
        let paginator = Entity::find()
            .filter(Column::Hide.eq("n"))
            .filter(Column::Type.eq("blog"))
            .order_by_desc(Column::Date)
            .paginate(&*self.db, per_page);
        let total = paginator.num_items().await?;
        let items = paginator.fetch_page(page.saturating_sub(1)).await?;
        Ok((items, total))
    }

    pub async fn find_by_category(&self, category_id: i32, page: u64, per_page: u64) -> Result<(Vec<Model>, u64), DbErr> {
        let paginator = Entity::find()
            .filter(Column::Sortid.eq(category_id))
            .filter(Column::Hide.eq("n"))
            .order_by_desc(Column::Date)
            .paginate(&*self.db, per_page);
        let total = paginator.num_items().await?;
        let items = paginator.fetch_page(page.saturating_sub(1)).await?;
        Ok((items, total))
    }

    pub async fn search(&self, keyword: &str, page: u64, per_page: u64) -> Result<(Vec<Model>, u64), DbErr> {
        let paginator = Entity::find()
            .filter(
                Condition::any()
                    .add(Column::Title.contains(keyword))
                    .add(Column::Content.contains(keyword))
            )
            .filter(Column::Hide.eq("n"))
            .order_by_desc(Column::Date)
            .paginate(&*self.db, per_page);
        let total = paginator.num_items().await?;
        let items = paginator.fetch_page(page.saturating_sub(1)).await?;
        Ok((items, total))
    }

    pub async fn create(&self, data: ActiveModel) -> Result<Model, DbErr> {
        data.insert(&*self.db).await
    }

    pub async fn update(&self, id: i32, data: ActiveModel) -> Result<Option<Model>, DbErr> {
        let model = Entity::find_by_id(id).one(&*self.db).await?;
        match model {
            Some(m) => {
                let mut active: ActiveModel = m.into();
                apply_if_set!(data, active,
                    title, content, excerpt, sortid, cover, hide, top, allow_remark, password
                );
                Ok(Some(active.update(&*self.db).await?))
            }
            None => Ok(None),
        }
    }

    pub async fn delete(&self, id: i32) -> Result<bool, DbErr> {
        let result = Entity::delete_by_id(id).exec(&*self.db).await?;
        Ok(result.rows_affected > 0)
    }

    pub async fn increment_views(&self, id: i32) -> Result<bool, DbErr> {
        let result = Entity::update_many()
            .col_expr(Column::Views, Expr::col(Column::Views).add(1))
            .filter(Column::Gid.eq(id))
            .exec(&*self.db)
            .await?;
        Ok(result.rows_affected > 0)
    }

    pub async fn count(&self) -> Result<u64, DbErr> {
        Entity::find().count(&*self.db).await
    }

    pub async fn filter(
        &self,
        category_id: Option<i32>,
        keyword: Option<&str>,
        status: Option<&str>,
        order: Option<&str>,
        page: u64,
        per_page: u64,
    ) -> Result<(Vec<Model>, u64), DbErr> {
        let mut query = Entity::find();

        if let Some(cat_id) = category_id {
            query = query.filter(Column::Sortid.eq(cat_id));
        }

        if let Some(kw) = keyword.filter(|k| !k.is_empty()) {
            query = query.filter(
                Condition::any()
                    .add(Column::Title.contains(kw))
                    .add(Column::Content.contains(kw))
            );
        }

        if let Some(st) = status.filter(|&s| s != "all") {
            query = query.filter(Column::Hide.eq(st));
        }

        query = match order {
            Some("views") => query.order_by_desc(Column::Views),
            Some("title") => query.order_by_asc(Column::Title),
            Some("date_asc") => query.order_by_asc(Column::Date),
            _ => query.order_by_desc(Column::Date),
        };

        let paginator = query.paginate(&*self.db, per_page);
        let total = paginator.num_items().await?;
        let items = paginator.fetch_page(page.saturating_sub(1)).await?;
        Ok((items, total))
    }
}
